pennylane/devices/default_qubit.py

```diff
@@ -436,6 +446,37 @@ def _apply_phase(self, state, axes, parameters, inverse=False):
         phase = self._conj(parameters) if inverse else parameters
         return self._stack([state[sl_0], phase * state[sl_1]], axis=axes[0])
 
+    def expval(self, observable, shot_range=None, bin_size=None):
+        """Returns the expectation value of a Hamiltonian observable. When the observable is a
+         `SparseHamiltonian` object, the expectation value is computed directly, which leads to
+          faster execution.
+
+        Args:
+            observable (~.Observable): a PennyLane observable
+            shot_range (tuple[int]): 2-tuple of integers specifying the range of samples
+                to use. If not specified, all samples are used.
+            bin_size (int): Divides the shot range into bins of size ``bin_size``, and
+                returns the measurement statistic separately over each bin. If not
+                provided, the entire shot range is treated as a single bin.
+
+        Returns:
+            float[real]/callable: returns the expectation value of the observable if it is a sparse
+            matrix in scipy coordinate list (COO) format or returns the :func:`~.expval` function
+            from the parent class
+        """
+        if observable.name == "SparseHamiltonian" and self.shots is None:
+
+            ev = coo_matrix.dot(
+                coo_matrix(self.state),
+                coo_matrix.dot(
+                    observable.matrix, coo_matrix(self.state.reshape(len(self.state), 1))
+                ),
+            )
+
+            return np.real(ev.toarray()[0])
+
+        return super().expval(observable, shot_range=shot_range, bin_size=bin_size)
```

Comment:
I prefer the first one for now since converting a large sparse matrix to a dense one is very memory-demanding. I am not sure if there is any advantage to work with a matrix representation of the Hamiltonian if it is not a sparse object. Any thoughts on this @josh146?

Reply:
> I am not sure if there is any advantage to work with a matrix representation of the Hamiltonian if it is not a sparse object. Any thoughts on this?

Yes, this is a good point. I suppose an alternative is to convert the sparse Hamiltonian COO representation into a linear combination of Pauli operators, which PL should then understand how to use?

> Thinking that the QNode class might be the right place to do it. Do you have any other suggestion?

I agree, this is the best place for it.

Reply:
Thanks Josh! Great point identifying the shots kwarg.

Let's just raise an error